pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java

```diff
@@ -79,28 +80,23 @@ private void attachTag(FlowNode currentNode, String syntheticContext) {
         }
     }
 
-    public static class GraphListenerAction extends InvisibleAction implements RunAction2 {
+    @Extension
+    public static class FlowExecutionListenerImpl extends FlowExecutionListener {
         @Override
-        public void onLoad(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        @Override
-        public void onAttached(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        private void attachListener(WorkflowRun run) {
-            if (run != null) {
-                FlowExecution exec = run.getExecution();
-                if (exec != null && !exec.isComplete()) {
-                    exec.addListener(new SyntheticStageGraphListener());
+        public void onRunning(FlowExecution execution) {
+            if (execution != null && !execution.isComplete()) {
+                // Only attach the graph listener if we know the run this execution is for is a Declarative run,
+                // i.e., if it has an ExecutionModelAction.
+                try {
+                    FlowExecutionOwner owner = execution.getOwner();
+                    if (owner != null && owner.getExecutable() instanceof WorkflowRun) {
+                        WorkflowRun r = (WorkflowRun)owner.getExecutable();
+                        if (r.getAction(ExecutionModelAction.class) != null) {
+                            execution.addListener(new SyntheticStageGraphListener());
```

Comment:
I will point out that this is actually more code than simply making `SyntheticStageGraphListener` an `@Extension` and having its `onNewHead` check whether its `owner` has `ExecutionModelAction`. (I suppose you could cache that calculation with a `CacheBuilder` with weak keys, though it should be very cheap anyway since it involves no I/O or significant computation.)